[RFC] vim-patch:7.4.{1648,1679,1681}

jbradaric opened the pull request:
```
vim-patch:7.4.1648

Problem:    Compiler has a problem copying a string into di_key[]. (Yegappan
            Lakshmanan)
Solution:   Add dictitem16_T.
```

https://github.com/vim/vim/commit/bee6c0cf86a75faa2aa893f2c9db82fd944a89a5

```
vim-patch:7.4.1679

Problem:    Coverity: copying value of v_lock without initializing it.
Solution:   Init v_lock in rettv_list_alloc() and rettv_dict_alloc().
```

https://github.com/vim/vim/commit/7d2a5796d39905a972e8f74af5f7b0a62e3de173

```
vim-patch:7.4.1681

Problem:    Coverity warns for fixed size buffer length (false positive).
Solution:   Add a check for the name length.
```

https://github.com/vim/vim/commit/ef9d9b94a8803c405884bb6914ed745ede57c596

jbradaric committed:
vim-patch:7.4.1679

Problem:    Coverity: copying value of v_lock without initializing it.
Solution:   Init v_lock in rettv_list_alloc() and rettv_dict_alloc().

https://github.com/vim/vim/commit/7d2a5796d39905a972e8f74af5f7b0a62e3de173

jbradaric committed:
vim-patch:7.4.1681

Problem:    Coverity warns for fixed size buffer length (false positive).
Solution:   Add a check for the name length.

https://github.com/vim/vim/commit/ef9d9b94a8803c405884bb6914ed745ede57c596

oni-link commented on src/nvim/eval.c:
```diff
@@ -4948,7 +4952,8 @@ static list_T *rettv_list_alloc(typval_T *rettv)
   rettv->vval.v_list = l;
   rettv->v_type = VAR_LIST;
-  ++l->lv_refcount;
+  rettv->v_lock = 0;
```
`VAR_UNLOCKED` instead of `0`

jbradaric replied:
@oni-link Thanks, I've changed the code to use VAR_UNLOCKED in both places.

oni-link commented on src/nvim/eval.c:
```diff
@@ -6128,7 +6133,8 @@ static void rettv_dict_alloc(typval_T *rettv)
   rettv->vval.v_dict = d;
   rettv->v_type = VAR_DICT;
-  ++d->dv_refcount;
+  rettv->v_lock = 0;
```
`VAR_UNLOCKED` instead of `0`

oni-link commented on src/nvim/eval.c:
```diff
@@ -514,6 +514,10 @@ void eval_init(void)
   for (size_t i = 0; i < ARRAY_SIZE(vimvars); i++) {
     p = &vimvars[i];
+    if (STRLEN(p->vv_name) > 16) {
+      EMSG("INTERNAL: name too long, increase size of dictitem16_T");
```
We don't have `dictitem16_T`.  That was introduced with vim patch 7.4.1648.

jbradaric replied:
@oni-link I've also added patch 7.4.1648 to the pull request.

oni-link commented on src/nvim/eval.c:
```diff
@@ -514,6 +514,10 @@ void eval_init(void)
     p = &vimvars[i];
+    if (STRLEN(p->vv_name) > 16) {
+      EMSG("INTERNAL: name too long, increase size of dictitem16_T");
+      getout(1);
```
Calling `getout()` here is not safe. For example pointer `p_mouse` is dereferenced, but it is only initialized after `eval_init()` (in `set_init_1()`).

If a name is too long:
```c
Program received signal SIGSEGV, Segmentation fault.
0x0000000000556f65 in setmouse () at /home/oni-link/git/neovim/src/nvim/mouse.c:461
461	  if (*p_mouse == NUL)
(gdb) bt
#0  0x0000000000556f65 in setmouse () at /home/oni-link/git/neovim/src/nvim/mouse.c:461
#1  0x0000000000549660 in wait_return (redraw=0) at /home/oni-link/git/neovim/src/nvim/message.c:782
#2  0x0000000000525ee2 in getout (exitval=1) at /home/oni-link/git/neovim/src/nvim/main.c:616
#3  0x0000000000472400 in eval_init () at /home/oni-link/git/neovim/src/nvim/eval.c:520
#4  0x0000000000525340 in early_init () at /home/oni-link/git/neovim/src/nvim/main.c:181
#5  0x0000000000525431 in main (argc=1, argv=0x7fffffffd918) at /home/oni-link/git/neovim/src/nvim/main.c:237
```
If name length is not a problem:
```c
(gdb) watch p_mouse
Hardware watchpoint 1: p_mouse
(gdb) r
Starting program: /home/oni-link/git/neovim/build/bin/nvim 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Hardware watchpoint 1: p_mouse

Old value = (char_u *) 0x0
New value = (char_u *) 0x6fd3b2 "a"
set_option_default (opt_idx=185, opt_flags=0, compatible=0) at /home/oni-link/git/neovim/src/nvim/option.c:825
825	        options[opt_idx].flags &= ~P_ALLOCED;
(gdb) bt
#0  set_option_default (opt_idx=185, opt_flags=0, compatible=0) at /home/oni-link/git/neovim/src/nvim/option.c:825
#1  0x00000000005828f4 in set_options_default (opt_flags=0) at /home/oni-link/git/neovim/src/nvim/option.c:868
#2  0x0000000000582336 in set_init_1 () at /home/oni-link/git/neovim/src/nvim/option.c:702
#3  0x0000000000525380 in early_init () at /home/oni-link/git/neovim/src/nvim/main.c:210
#4  0x0000000000525431 in main (argc=1, argv=0x7fffffffd918) at /home/oni-link/git/neovim/src/nvim/main.c:237
```

fwalch replied:
@oni-link Do you know if this also (still?) a problem in Vim?

jbradaric replied:
I've just tried it and Vim also segfaults if a variable name is longer than 16 characters.

jamessan replied:
Does it really matter?  This code is just here to flag an issue when a new item is added to the vimvars array.  The entire point is that you're supposed to notice that the variable name is too long and the array size should be changed.

Why not just change the if to `assert(STRLEN(p->vv_name) <= 16);`?  Then the getout and EMSG aren't needed.

justinmk replied:
@jamessan Agreed, but how about `abort()` so that Release builds are protected as well. `assert()` should be for things that "cannot happen".

jamessan replied:
I figured assert would be good enough because CI should catch the problem before it gets to users.

jbradaric replied:
@jamessan I've replaced the if with an `assert`, per your suggestion.

jbradaric committed:
vim-patch:7.4.1648

Problem:    Compiler has a problem copying a string into di_key[]. (Yegappan
            Lakshmanan)
Solution:   Add dictitem16_T.

https://github.com/vim/vim/commit/bee6c0cf86a75faa2aa893f2c9db82fd944a89a5
